'New' approach for Tuya MCU Cluster

javicalle opened the pull request:
Here I come with another approach for (all?) Tuya clusters that send commands through the MCU.
This approach tries to extend the `TuyaNewManufCluster` by also implementing Device --> MCU comunication, that is, sending commands to the DP.

I have taken the opportunity to segregate the MCU related implementations to a new folder. The idea would be for the rest of the classes to be migrated.

My starting point has been issue #1233 which allows me to release the new implementation without the risk of affecting operational devices.

I will try to explain the new implementation and the pros and cons that I have found.
- segregated classes that should help maintain functionality
- take advantage of the power of the `TuyaNewManufCluster` class. Its declarative model may not be intuitive but I think it has potential to be leveraged in more implementations. All the credit to its author (who is not me)
- maintains the use of a Bus to communicate `TUYA_MCU_COMMAND` events to the Tuya cluster. Its use has been discussed in #1153 but I haven't found a better solution
- new approaches to Tuya clusters (explained below)
- support for writing attributes (not implemented yet). It will allow modifying device attributes such as the `backlight_mode` or the `power_on_state`
- `TUYA_MCU_VERSION_RSP` message supported. New `mcu_version` attribute added to the `TuyaMCUCluster`. Reported version can be read in raw and `x.y.z` format.



The new approach to Tuya clusters makes use of the `TuyaCommad` classes instead of the `TuyaManufCluster.Command` classes. For my use cases they have been comfortable for me to transform the values.
Now the all clusters send a `TuyaClusterData` to the Bus (and not values). Each cluster knows how manage its values (is it a Bool, a String or an Enum?) And the `TuyaMCUCluster` will be only in charge of communicating with the MCU.


Some of the improvement points:
* [ ] `get_dp_from_cluster` method. Performs a reverse search to find the appropriate DP. Not the most elegant one.
* [x] ~~`cmd_payload = TuyaCommand()` flow. Static and repetitive. I thought I saw somewhere where it initializes on a line or two, but I'm not sure. Candidate to take it out to a helper class builder from the `TuyaDPType` and the value.~~
* [ ] response to sending commands and writing attributes is now decoupled from client (due to the use of the `command_bus`). This can mask errors and problems for the lack of response.
* [ ] Cluster attributes have been overwritten to leave only the ones that are actually handled by the MCU. It's probably wise to do the same for cluster commands.


The code has been tested locally with a `TuyaDoubleSwitch` device, but the device has not been included yet due to the risk of affecting multiple devices.
The included device implementation `TuyaDoubleSwitchDimmer` I have NOT been able to test it as I do not have a device, so the code will need to be validated before doing the merge.

(Fixes: #1033)

MattWestb commented:
I have asking the core devs before but no have answering the question if the grup and scene cluster have some user in the Zigbee module used for DP / MCU commands.
By Zigbee standard one device must have commissioning (0x0000) and basic cluster and also identify im many device types.
The old modules have OTA cluster and it shall being left if we is getting one updat of the Zigbee module in the future.

The group and scene i cant finding any use of on the module and making user trying binding and adding groups to the device that is not working.

I have trying with my TRV deleting them and it was working OK but its need more testing and i dont have more DP devices so i cant saying it working on all of them.
I think if its working on the old (TYST11) and the new (TZE200) its good deleting  them in the replace so its not making problem with user and system in the future.

Can you trying then you have time if your switches is working OK if deleting all not used cluster ?
If OK i think then converting all dp devices to the new MCU cluster we can rolling it on all quirk so getting it more clean and looking the same all over.

javicalle commented:
> Can you trying then you have time if your switches is working OK if deleting all not used cluster ?

Sure, I will do some test in the next days and report here.

javicalle commented:
After some tests and a few changes, this version must be fully operational.

Some changes:
* Implemented a new `TuyaDPType` class that helps in the Tuya <--> Zigbee data type transformation
* Implemented a new `MCUVersion` class
  * fixes the payload serialization
  * format version value to x.y.z
* Now `TUYA_MCU_VERSION_RSP` messages are managed and value are stored in a new `mcu_version` attribute (this value is updated only when device broadcast the message)
* `write_attributes` method generates the command payload dynamically.

javicalle commented:
I am trying to implement tests that are valid and functional.

Another point to comment/review would be the values used in the attributes (standard and custom).

In the Tuya cluster:
* as `attributtes`:
  * 0x8001: ("backlight_mode", t.enum8),
  * 0x8002: ("power_on_state", t.enum8),
* as `manufacturer_attributes`:
  * 0xEF00: ("mcu_version", t.uint48_t),

In the `TuyaOnOff` cluster:
* as `attributtes`:
  * 0x0000: ("on_off", t.Bool),

In the `TuyaLevelControl` cluster:
* as `attributtes`:
  * 0x0000: ("current_level", t.uint8_t)
* as `manufacturer_attributes`:
  * 0xEF01: ("minimum_level", t.uint8_t),
  * 0xEF02: ("bulb_type", t.enum8),

nworbneb commented:
I've updated to latest PR code and this is working as expected with Samotech (Moes) MS-105BZ dual dimmer.

dmulcahey commented:
@javicalle @MattWestb are you folks on discord?

MattWestb commented:
Im not and i cant do deep coding only copy / paste ;-((

javicalle commented:
Me neither, sorry

javicalle commented:
@dmulcahey  Something do you want to talk or discuss?
I can try to use Discord if necessary.

javicalle commented:
Another iteration over the code.

ZB --> DP converter capability has been added. This is needed in the conversion from ZB attributes to DP values.
Now the `write_attributes`part must be fully functional.

I'm going to try to get someone to validate the `TuyaDoubleSwitchDimmer` implementation, but I don't think it needs to be modified other than to fix a bug.

In parallel, I want to check if it adds anything to any of the existing implementations.

dmulcahey commented:
> @dmulcahey Something do you want to talk or discuss?
> I can try to use Discord if necessary.

I have the same handle on discord. If it isn't too much of a pain can you dm me when you have a moment?

dmulcahey commented on tests/test_tuya_dimmer.py:
```diff
@@ -0,0 +1,256 @@
+    assert dimmer2_listener.attribute_updates[1][0] == 0x0000
+    assert dimmer2_listener.attribute_updates[1][1] == 170
+
+    # # events from channel 2 updates only EP 2
```
These should be uncommented and made to work or they should be removed from the PR.

javicalle replied:
Done

Tropaion commented:
I don't understand much how it works, but I'm recently got many zigbee devices so I'm interested in the progress here or if I should try Zigbee2MQTT. Does it work already? How long does it normally take for it to get into HA (stable)?
And is it possible to disable/dimm the button background light (for bedrooms)?

javicalle commented:
Now the code it's ready to merge.

Most of the existing classes involved have been extended or overwritten. My intention is to migrate the use of the old classes to the new ones and eliminate duplicate code in next versions, but not a big bang for now.

nworbneb commented:
Tested latest version with HA 2021.12.10 with a Samotech MS-105BZ and all still working as expected.  Both gangs work with switching, dimming and I can configure minimum dimming level.  Thanks for your hard work @javicalle, let's get this in 2022.2!

Tropaion commented:
Does this only affect the tuya dimmer or other devices too?
I'm currently using a few different tuya devices, but they don't work that well currently so i'm deliberating if I should switch to Zigbee2MQTT

javicalle commented:
> Does this only affect the tuya dimmer or other devices too? I'm currently using a few different tuya devices, but they don't work that well currently so i'm deliberating if I should switch to Zigbee2MQTT

This approach is intended to solve the problem with "multi-endpoint" devices. Most devices with a single endpoint have been implemented (with more or less success) but devices that required more than one endpoint did not have an easy fit.

Being a new approach (we could say experimental), for now it has only been implemented for `_TZE200_e3oitdyu` (multidimmer switch).

Maybe the approach can be adapted for other devices, but not sure if this is a better approach o worse to the current implementations. That is something that the maintainers of current devices should consider.

If your problems are related to some unsupported or multi endpoint device, I can try to see if it's possible to generate a quirk. But I'm sorry to say that if someone has already quirked the device, I have no plans to replace it with a new implementation. I hope you can understand my position.

Regards.

dmulcahey commented on zhaquirks/tuya/mcu/__init__.py:
```diff
@@ -0,0 +1,462 @@
+            self.debug("raw: %s", cmd_payload.data.raw)
+            return cmd_payload
+        else:
+            self.info(
```
can you make this a warning

dmulcahey commented on zhaquirks/tuya/mcu/__init__.py:
```diff
@@ -0,0 +1,462 @@
+                TuyaMCUCluster.ep_attribute,
+                "power_on_state",
+                dp_type=TuyaDPType.ENUM,
+                # lambda x: PowerOnState(x),
```
stale?

dmulcahey commented on zhaquirks/tuya/mcu/__init__.py:
```diff
@@ -0,0 +1,462 @@
+                TuyaMCUCluster.ep_attribute,
+                "backlight_mode",
+                dp_type=TuyaDPType.ENUM,
+                # lambda x: BackLight(x),
```
stale?

dmulcahey commented on tests/test_tuya_mcu.py:
```diff
@@ -0,0 +1,126 @@
+
+zhaquirks.setup()
+
+# ZCL_TUYA_VERSION_RSP = b"\tp\x17\x00\x03\x82"
```
stale?

dmulcahey commented:
few nit type comments. Really nice work. Thanks for the contribution!

javicalle commented:
Thanks. @dmulcahey, @MattWestb and all testers.

Tropaion commented:
I also have a TS0601 but the quirk needs to be modified a bit which I wanted to try now.
```
{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.AllocateAddress|RxOnWhenIdle|MainsPowered|FullFunctionDevice: 142>, manufacturer_code=4417, maximum_buffer_size=66, maximum_incoming_transfer_size=66, server_mask=10752, maximum_outgoing_transfer_size=66, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": 260,
      "device_type": "0x0051",
      "in_clusters": [
        "0x0000",
        "0x0004",
        "0x0005",
        "0xef00"
      ],
      "out_clusters": [
        "0x000a",
        "0x0019"
      ]
    },
    "242": {
      "profile_id": 41440,
      "device_type": "0x0061",
      "in_clusters": [],
      "out_clusters": [
        "0x0021"
      ]
    }
  },
  "manufacturer": "_TZE200_3p5ydos3",
  "model": "TS0601",
  "class": "zigpy.device.Device"
}
```
But I don't know how to add the new MCU cluster for tests, can someone tell me? 😄

javicalle commented:
@Tropaion sure!
Please, follow up in #1302 where are working with this kind of devices.

Tropaion commented:
@javicalle Its probably some minor mistake on my side but I can't get the import working, i renamed the file and tried:
```from ./mcu.py import...```
```from .mcu import...```
```from /config/custom_zha_quirks/mcu.py import...```
But I always get an error:
```
 File "/config/custom_zha_quirks/ts0601_dimmer.py", line 21
    from /config/custom_zha_quirks/mcu.py import (
         ^
SyntaxError: invalid syntax
```

javicalle commented:
As posted here https://github.com/zigpy/zha-device-handlers/issues/1302#issuecomment-1019572315
try with:
```python
from custom_zha_quirks.tuya.mcu import (
    TuyaInWallLevelControl,
    TuyaLevelControlManufCluster,
    TuyaOnOff as TuyaOnOffMCU,
)
```

javicalle commented:
Sorry, but this issue is already closed.
Let's continue in #1302 where you can find an experimental quirk and the steps to test it:
* https://github.com/zigpy/zha-device-handlers/issues/1302#issuecomment-1019572315

Regards.
